### PR TITLE
Add test cases for subgrid scale wetting and drying corrections

### PR DESCRIPTION
**Pull request by sbrus89**
This pull request introduces two test cases for the subgrid scale correction scheme added in this E3SM PR: https://github.com/E3SM-Project/E3SM/pull/6288. These test cases were developed under the ICoM project in collaboration with Dam Wirasaet at U. of Notre Dame.

The subgrid corrections used in these tests account for the nonlinear relationship between ssh and grid-averaged wet volume per unit area for partially wet cells. These relationships are calculated as lookup tables from high-resolution, subgrid scale, bathymetry data. This pre-processing step is implemented in MPAS-Ocean init mode :see_no_evil:. Using the subgrid correction scheme increases the accuracy of wetting and drying, and is especially advantageous at coarser resolutions. Currently, this capability is only implemented for single layer configurations.

The two test cases are:
1. A subgrid version of the existing parabolic bowl test case

2. A simulation of Buttermilk Bay, MA with idealized tidal boundary forcing

More information on the subgrid correction methodology can be found in [Kennedy et al. 2019](https://doi.org/10.1016/j.ocemod.2019.101491). This paper also discusses the Buttermilk Bay test case.

This PR also exercises a new (optional) update to the existing wetting and drying algorithm that takes into account the ssh gradient between cells that share a previously limited edge. This treatment of limited edges improves the wetting and drying convergence for the parabolic bowl. However, this approach is currently limited to a single layer configuration. Thanks to Dam for contributing this improvement!

Checklist
* [x] User's Guide has been updated
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.rst`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://mpas-dev.github.io/compass/latest/developers_guide/building_docs.html) and changes look as expected
* [ ] The `E3SM-Project` submodule has been updated with relevant E3SM changes
* [x] Document (in a comment titled `Testing` in this PR) any testing that was used to verify the changes

### CODE REVIEW

**Comment by sbrus89** [comment]
### Testing

These are the results of the parabolic bowl subgrid case:
points
solution_360

The standard wetting and drying results for the parabolic bowl:
points
solution_360

The convergence comparison between the subgrid and standard cases:
error

For comparison, the convergence plot without the limited edge ssh treatment is:
error

**Comment by mark-petersen** [comment]
Thanks! Does this file need to be added?
```
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://web.lcrc.anl.gov/public/e3sm/mpas_standalonedata/mpas-ocean/initial_condition_database/buttermilk_bathy.nc
```

**Comment by xylar** [comment]
@sbrus89, bathymetry would ideally go in the bathymetry database, not in the initial condition database:
https://web.lcrc.anl.gov/public/e3sm/mpas_standalonedata/mpas-ocean/bathymetry_database/
Picky, I know...

**Comment by xylar** [comment]
And sorry for not reviewing.  I was waiting on the E3SM PR.

**Comment by xylar** [comment]
Thanks!

**Comment by sbrus89** [comment]
@mark-petersen, the bathymetry should be available now.

**Comment by mark-petersen** [comment]
Passes all tests using intel on chrysalis and gnu on perlmutter, both debug and optimized. Tests are as follows:

**perlmutter**
```
./conda/configure_compass_env.py --conda ~/miconda3 --compiler gnu -m pm-cpu
source load_dev_compass_1.2.0-alpha.9_pm-cpu_gnu_mpich.sh
compass list|grep -i parabolic_bowl
 370: ocean/parabolic_bowl/standard/ramp_lts
 371: ocean/parabolic_bowl/standard/ramp
 372: ocean/parabolic_bowl/standard/noramp_lts
 373: ocean/parabolic_bowl/standard/noramp
 374: ocean/parabolic_bowl/subgrid/ramp
 375: ocean/parabolic_bowl/subgrid/noramp

compass list|grep -i butter
 107: ocean/buttermilk_bay/standard
 108: ocean/buttermilk_bay/subgrid

compass setup -p ~/repos/E3SM/pr/components/mpas-ocean -w $r/230315_subgrid_sbrus_opt -n 107 108 374 375

salloc --nodes 1 --qos interactive --time 01:00:00 --constraint cpu --account=e3sm
source load_dev_compass_1.2.0-alpha.9_pm-cpu_gnu_mpich.sh
cd  $r/230315_subgrid_sbrus_opt
compass run
```

**chyrsalis**
```
./conda/configure_compass_env.py --conda ~/miconda3 --compiler intel
source load_dev_compass_1.2.0-alpha.9_chrysalis_intel_openmpi.sh
compass setup -p ~/repos/E3SM/pr/components/mpas-ocean -w $r/230315_subgrid_sbrus_opt -n 107 108 374 375

srun -p debug -N 1 -t 1:00:00 --pty bash
source load_dev_compass_1.2.0-alpha.9_pm-cpu_gnu_mpich.sh
cd  $r/230315_subgrid_sbrus_opt
compass run
```

**Review by mark-petersen — APPROVED** [review]
@sbrus89 this is excellent! Approving based on documentation, output plots, and passing functional testing with two compilers. I did not review the details of the compass code implementation.

**Comment by sbrus89** [comment]
Thanks for testing, @mark-petersen!

File: compass/ocean/tests/buttermilk_bay/default/__init__.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        Parameters
+        ----------
+        test_group : compass.ocean.tests.buttermilk_bay.ButtermilkBay
+            The test group that this test case belongs to
+
+        wetdry : str
+            The type of wetting and drying used (``standard``, ``subgrid``)
+        """
+        name = f'{wetdry}'
+        subdir = f'{wetdry}'
```
Comment (cbegeman) [inline compass/ocean/tests/buttermilk_bay/default/__init__.py:34]:
```suggestion
        name = wetdry
        subdir = wetdry
```

File: compass/ocean/tests/buttermilk_bay/default/__init__.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+        goal_cells_per_core = config.getfloat('buttermilk_bay',
+                                              'goal_cells_per_core')
+        max_cells_per_core = config.getfloat('buttermilk_bay',
+                                             'max_cells_per_core')
+
+        for resolution in self.resolutions:
+
+            lx = config.getfloat('buttermilk_bay', 'Lx')
+            ly = config.getfloat('buttermilk_bay', 'Ly')
```
Comment (cbegeman) [inline compass/ocean/tests/buttermilk_bay/default/__init__.py:69]:
```suggestion
        lx = config.getfloat('buttermilk_bay', 'Lx')
        ly = config.getfloat('buttermilk_bay', 'Ly')

            for resolution in self.resolutions:
```

File: compass/ocean/tests/buttermilk_bay/forward.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+import time
+
+from compass.model import run_model
+from compass.step import Step
+
+
+class Forward(Step):
+    """
+    A step for performing forward MPAS-Ocean runs as part of parabolic bowl
```
Comment (cbegeman) [inline compass/ocean/tests/buttermilk_bay/forward.py:9]:
```suggestion
    A step for performing forward MPAS-Ocean runs as part of buttermilk bay
```

File: docs/users_guide/ocean/test_groups/buttermilk_bay.rst [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+==============
+
+The ``buttermilk_bay`` test group implements a realistic regional
+simulation of tidally forced circulation in a multi-bay system 
+connected by narrow channels. The computational domain is 
+`Buttermilk Bay, Massachusetts <https://tinyurl.com/4zpz36ft>`_.
+
+.. _ocean_buttermilk_bay_default:
+
+default
```
Comment (cbegeman) [inline docs/users_guide/ocean/test_groups/buttermilk_bay.rst:13]: Maybe state somewhere that the ramp feature is used?

**Review by cbegeman — APPROVED** [review]
This looks fantastic! I ran the standard test cases successfully. Just a few minor suggestions.

File: compass/ocean/tests/buttermilk_bay/default/__init__.py (hunk 1 above)
Comment (sbrus89) [reply compass/ocean/tests/buttermilk_bay/default/__init__.py:34]: Thanks for noticing this!

File: compass/ocean/tests/buttermilk_bay/default/__init__.py (hunk 2 above)
Comment (sbrus89) [reply compass/ocean/tests/buttermilk_bay/default/__init__.py:69]: Good catch, thanks.

File: compass/ocean/tests/buttermilk_bay/forward.py (hunk 3 above)
Comment (sbrus89) [reply compass/ocean/tests/buttermilk_bay/forward.py:9]: Ugh, I thought I caught all these!

File: docs/users_guide/ocean/test_groups/buttermilk_bay.rst (hunk 4 above)
Comment (sbrus89) [reply docs/users_guide/ocean/test_groups/buttermilk_bay.rst:13]: Good idea, will do.

**Comment by xylar** [comment]
#821 will update the E3SM-Project submodule with the changes needed to support this PR.  @sbrus89, could you rebase this branch once that goes in?

**Comment by sbrus89** [comment]
@xylar, thanks for the submodule update. I just rebased.

**Comment by xylar** [comment]
@sbrus89, I was able to run the Buttermilk Bay test just fine!

I had trouble with the viz step in both parabolic bowl test cases.  The `ramp` test produces:
```
  in /gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/ocean/tests/parabolic_bowl/viz/__init__.py

      Failed
Exception raised while running the steps of the test case
Traceback (most recent call last):
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/run/serial.py", line 320, in _log_and_run_test
    _run_test(test_case, available_resources)
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/run/serial.py", line 417, in _run_test
    _run_step(test_case, step, test_case.new_step_log_file,
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/run/serial.py", line 468, in _run_step
    step.run()
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/ocean/tests/parabolic_bowl/viz/__init__.py", line 50, in run
    self.rmse_plots()
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/ocean/tests/parabolic_bowl/viz/__init__.py", line 227, in rmse_plots
    rmse[i] = self.compute_rmse(
              ^^^^^^^^^^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/ocean/tests/parabolic_bowl/viz/__init__.py", line 262, in compute_rmse
    ind = time.index(dt.datetime.strptime('0001-01-03_18:00:00',
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: datetime.datetime(1, 1, 3, 18, 0) is not in list
```

The `noramp` test leads to:
```
Exception raised while running the steps of the test case
Traceback (most recent call last):
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/run/serial.py", line 320, in _log_and_run_test
    _run_test(test_case, available_resources)
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/run/serial.py", line 417, in _run_test
    _run_step(test_case, step, test_case.new_step_log_file,
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/run/serial.py", line 468, in _run_step
    step.run()
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/ocean/tests/parabolic_bowl/viz/__init__.py", line 49, in run
    self.contour_plots(points)
  File "/gpfs/fs1/home/ac.xylar/compass/subgrid_pr/compass/ocean/tests/parabolic_bowl/viz/__init__.py", line 164, in contour_plots
    ds['ssh'][i, :],
    ~~~~~~~~~^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/dataarray.py", line 874, in __getitem__
    return self.isel(indexers=self._item_key_to_dict(key))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/dataarray.py", line 1493, in isel
    variable = self._variable.isel(indexers, missing_dims=missing_dims)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/variable.py", line 1018, in isel
    return self[key]
           ~~~~^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/variable.py", line 785, in __getitem__
    data = indexing.apply_indexer(indexable, indexer)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/indexing.py", line 1002, in apply_indexer
    return indexable[indexer]
           ~~~~~~~~~^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/indexing.py", line 820, in __getitem__
    return type(self)(_wrap_numpy_scalars(self.array[indexer]))
                                          ~~~~~~~~~~^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/indexing.py", line 770, in __getitem__
    return type(self)(_wrap_numpy_scalars(self.array[indexer]))
                                          ~~~~~~~~~~^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/indexing.py", line 645, in __getitem__
    return type(self)(self.array, self._updated_key(indexer))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/indexing.py", line 600, in _updated_key
    full_key.append(_index_indexer_1d(k, next(iter_new_key), size))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/fs1/home/ac.xylar/chrysalis/miniforge3/envs/compass_test2/lib/python3.11/site-packages/xarray/core/indexing.py", line 287, in _index_indexer_1d
    indexer = _expand_slice(old_indexer, size)[applied_indexer]  # type: ignore[assignment]
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
IndexError: index 80 is out of bounds for axis 0 with size 79
```

**Comment by xylar** [comment]
Could you see if you are able to reproduce those errors?  (This was on Chrysalis with the default Intel and OpenMPI configuration.)

**Comment by sbrus89** [comment]
@xylar, thanks for testing. I'm looking into it now.

**Comment by sbrus89** [comment]
@xylar, both the `ocean/parabolic_bowl/subgrid/ramp` and `ocean/parabolic_bowl/subgrid/noramp` seemed to pass for me on Chrysalis with the Intel/OpenMPI config. Can you point me to your work directory so I can look more closely at what the issue might be?

**Comment by xylar** [comment]
@sbrus89, I'll try running again first and point to the errors if they happen again.

**Review by xylar — APPROVED** [review]
@sbrus89, I reran and didn't see the errors from before.  I must have messed things up the first time.  My job ran out of time and maybe stepped on its own toes somehow.

**Comment by sbrus89** [comment]
Great, thank you @xylar!

File: compass/ocean/tests/parabolic_bowl/viz/__init__.py [hunk 5]
```diff
@@ -233,7 +239,7 @@ def rmse_plots(self):
         for i in range(len(self.resolutions) - 1):
             rmse_1st_order[i + 1] = rmse_1st_order[i] / 2.0
 
-        ax.loglog(self.resolutions, np.flip(rmse_1st_order),
+        ax.loglog(self.resolutions, rmse_1st_order,
```
Comment (sbrus89) [inline compass/ocean/tests/parabolic_bowl/viz/__init__.py:242]: @cbegeman, I forgot I made this change in this PR related to #765 .

Comment (cbegeman) [reply compass/ocean/tests/parabolic_bowl/viz/__init__.py:242]: @sbrus89 Great. I just closed the issue.

**Pull request merged**